lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,185 @@
+
+        keys = ["duration", "packets", "frags", "errors"]
+
+        try:
+            results = dict(  # converts iterator to actual dict
+                zip(  # zips values with `keys`
+                    keys, map(int, match.groups())  # map converts str numbers to int
+                )
+            )
+        except ValueError:  # just to be sure all the nums can be converted to int
+            raise TestModuleError("Could not convert numbers in pktgen results")
+
+        results["duration"] /= 1_000_000  # from usec to sec
+        return results
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    src_if = DeviceParam()
+    dst_mac = StrParam()
+
+    src_ip = IpParam()
+    dst_ip = IpParam()
+
+    count = IntParam(mandatory=False, default=0)  # 0 = no upper limit
+    size = IntParam(mandatory=False, default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(mandatory=False, default=1)
+    burst = IntParam(mandatory=False, default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self._devices = []
+
+        self._res_data = {}
+        self._output_parser = None
+
+    def run(self):
+        if not kmod_in_use("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
+
+        logging.debug("Configuring PktGen generator")
+        self._pg_ctl("reset")
+        self._configure_generator()
+
+        output_parser = PktGenResultsParser(self._devices, self._res_data)
+
+        logging.debug("Starting generator")
+        pktgen = Popen("echo 'start' > /proc/net/pktgen/pgctrl", shell=True)
+        # ^^ echoing start to controller is blocking => needs to be separated
+
+        time.sleep(self.params.duration)
+
+        pktgen.kill()  # stops pktgen
+
+        output_parser.parse_dev_outputs()
+
+        self._deconfigure_generator()
+        return True
+
+    def _configure_generator(self):
+        logging.debug("Configuring generator")
+
+        ipv6 = True
+        if isinstance(self.params.src_ip, Ip4Address):
+            ipv6 = False
+
+        src = f"src{6 if ipv6 else ''}"
+        dest = f"dst{6 if ipv6 else ''}"
+
+        for cpu in self.params.cpus:
+            dev = f"{self.params.src_if.name}@{cpu}"
+            logging.debug(f"Adding interface {self.params.src_if.name} to cpu {cpu}")
+
+            self._pg_thread(cpu, f"add_device {dev}")
+
+            self._pg_set(cpu, f"flag QUEUE_MAP_CPU")
+            self._pg_set(cpu, f"count {self.params.count}")
+            self._pg_set(cpu, f"pkt_size {self.params.size}")
```

Comment:
Hmm, what do you mean by parameters?

`flag`, `count`, `pkt_size` have the same names here as in the pktgen, it won't work otherwise.

The methods are also called the same as in bash version of pktgen, see https://github.com/torvalds/linux/blob/master/samples/pktgen/functions.sh .

Reply:
i think he means the rename of `self.params.size` to `self.params.pkt_size`

Reply:
Yes, I meant rename of `self.params.size` to `self.params.pkt_size`.

Reply:
Sure, done